### ISSUE TITLE
extra margin seems to come from "empty" sections...

### ISSUE THREAD

**Issue by williamnoto** [mikeswanson/JBSpacer#2]
I really like this library but I'm struggling with some extra margin that piles up when I have empty sections... Checkout the piled up margin in the screenshot below... I _believe_ those are from the internal margin padding of 3 empty sections.

![screen shot 2015-04-15 at 9 56 53 pm](https://cloud.githubusercontent.com/assets/6032433/7173287/6ce0f7cc-e3ba-11e4-89d4-3d3f9a52d09d.png)

in my viewDidLoad I invoke JBSpacer:

```
JBSpacerOption *option = [JBSpacerOption optionWithItemSize:[self itemWidth]
                                              minimumGutter:2.0f
                                        gutterToMarginRatio:1.0f
                                              availableSize:self.collectionView.frame.size.width
                                   distributeExtraToMargins:YES];
JBSpacer *spacer = [JBSpacer spacerWithOption:option];

[spacer applySpacingToCollectionViewFlowLayout:(UICollectionViewFlowLayout*)self.collectionViewLayout];
```

and later I have:

```
- (CGFloat) itemWidth {
    CGFloat margin = 5.0f;
    CGFloat collectionViewWidth = self.collectionView.frame.size.width;
    CGFloat width = (collectionViewWidth-(3*margin)) / 2;
    return width;
}

- (CGSize)collectionView:(UICollectionView *)collectionView
              layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    CGFloat length = [self itemWidth];
    return CGSizeMake(length, length); // width, height
}

- (UICollectionReusableView*) collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
     UICollectionReusableView *reusableview = nil;
    if (kind == UICollectionElementKindSectionHeader)
    {
        HNTileBrowsingHeaderViewCollectionReusableView * header = (HNTileBrowsingHeaderViewCollectionReusableView*)[collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:@"HeaderView" forIndexPath:indexPath];
        NSArray * sectionArray = _sectionedResults[indexPath.section];
        header.headerViewLabel.text = [_sectionHeaderLabels objectForKey:sectionArray];
        reusableview = header;
    }
    return reusableview;
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    NSArray * itemArray = _visibleSectionedResults[section];
    if (itemArray.count == 0)
        return CGSizeZero;
    CGSize size = CGSizeMake(320, (itemArray.count > 0 ? 30.0f : 1)); // width gets ignored.
    return size;
}
```

If I comment out the JBSpacer from viewDidLoad I get this:

![screen shot 2015-04-15 at 10 06 02 pm](https://cloud.githubusercontent.com/assets/6032433/7173360/b90efbd4-e3bb-11e4-8283-cbba333cec69.png)

See how that margin is gone from the empty sections (but now I don't have the nice layout in my collection? :(

**Comment by mikeswanson** [comment]
My guess is that you're correct that the insets are "piling up" in the empty sections. You may have to provide custom insets using the delegate for sections that are empty. This might help: http://stackoverflow.com/questions/13970950/uicollectionview-spacing-margins

**Comment by williamnoto** [comment]
Hi Mike, 

Thanks so much for your very prompt reply. That did seem to make a difference, except now I don’t seem to be getting the benefit of the spacer anymore. This probably sounds stupid (I’ve been up all day, I apologize) but can I capture the insets your tool defines so I can do something more nuanced than this:
- (UIEdgeInsets)collectionView:(UICollectionView _)collectionView layout:(UICollectionViewLayout_)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
  NSArray \* itemArray = _visibleSectionedResults[section];
  if (itemArray.count == 0)
      return UIEdgeInsetsMake(CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN);
  else
      return UIEdgeInsetsMake(2,2,2,2);
  }

that is, how do I replace those 2’s with the values you calculate…

**Comment by williamnoto** [comment]
well this is a big improvement…
- (UIEdgeInsets)collectionView:(UICollectionView _)collectionView layout:(UICollectionViewLayout_)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
  NSArray \* itemArray = _visibleSectionedResults[section];
  if (itemArray.count == 0)
      return UIEdgeInsetsMake(CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN);
  else {
      CGFloat margin = [_spacer spacing].margin;
      return UIEdgeInsetsMake(margin,margin,margin,margin);
  }
  }

**Comment by mikeswanson** [comment]
Looks like you're ahead of me. The only issue you'll have to deal with is the "extra" space (if you care). Another option may be to apply the results of the spacer to the layout, then actually use the layout values in your delegate method. That way, the spacing will have been taken care of.

**Comment by williamnoto** [comment]
hmm… which delegate method? sorry to be obtuse.

**Comment by mikeswanson** [comment]
The collectionView:layout:insetForSectionAtIndex: delegate method that you're now using. You should be able to access ((UICollectionViewFlowLayout *)collectionViewLayout).sectionInset to get the exact values that are set by JBSpacer.

**Comment by williamnoto** [comment]
Oh, man, I am really out of it. I forgot to show you that I was doing this:
- (UIEdgeInsets)collectionView:(UICollectionView _)collectionView layout:(UICollectionViewLayout_)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
  NSArray \* itemArray = _visibleSectionedResults[section];
  if (itemArray.count == 0)
      return UIEdgeInsetsMake(CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN);
  else {
      CGFloat margin = [_spacer spacing].margin;
      return UIEdgeInsetsMake(margin,margin,margin,margin);
  }
  }

and that _spacer object is set in viewDidLoad… just a property…

so viewDidLoad now has:

JBSpacerOption _option = [JBSpacerOption optionWithItemSize:[self itemWidth]
                                                  minimumGutter:2.0f
                                            gutterToMarginRatio:1.0f
                                                  availableSize:self.collectionView.frame.size.width
                                       distributeExtraToMargins:YES];
    _spacer = [JBSpacer spacerWithOption:option];
    [_spacer applySpacingToCollectionViewFlowLayout:(UICollectionViewFlowLayout_)self.collectionViewLayout];

That should do the same thing, right? Well, just double checked changing it to:
- (UIEdgeInsets)collectionView:(UICollectionView _)collectionView layout:(UICollectionViewLayout_)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
  NSArray \* itemArray = _visibleSectionedResults[section];
  if (itemArray.count == 0)
      return UIEdgeInsetsMake(CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN, CGFLOAT_MIN);
  else {
  //        return ((UICollectionViewFlowLayout *)collectionViewLayout).sectionInset;
      CGFloat margin = [_spacer spacing].margin;
      return UIEdgeInsetsMake(margin,margin,margin,margin);
  }
  }

and I think they look pretty much the same…

Really appreciate your digging in to this with me.

**Comment by mikeswanson** [comment]
Of course, I don't have the project to test, but it seems to me that you were close to having a solution that I would expect to work. You create the JBSpacer and use it to apply the global sectionInsets to your flow layout in viewDidLoad. That's all good. Then, you provide custom insets based on the insetForSectionAtIndex. My suggestion was to basically do what you've commented-out in your example: return the base flow layout's sectionInset property (which was set earlier by the JBSpacer in viewDidLoad) for any section that doesn't contain 0 items.

**Comment by williamnoto** [comment]
OK, great. Thanks again for all of your help. Really appreciate it.

**Comment by zl19850227** [comment]
JBSpacer can be used for this layou?
![-4](https://cloud.githubusercontent.com/assets/4365294/12161198/1597eba6-b52e-11e5-8e59-d49864749d59.png)
